pyvista/core/utilities/features.py

```diff
@@ -99,6 +99,120 @@ def voxelize(mesh, density=None, check_surface=True):
+
+    Examples
+    --------
+    Create an equal density voxel volume from input mesh.
+
+    >>> import pyvista as pv
+    >>> import numpy as np
+
+    Load file from PyVista examples.
+
+    >>> from pyvista import examples
+    >>> mesh = examples.download_cow()
+
+    Create an equal density voxel volume and plot the result.
+
+    >>> vox = pv.voxelize_volume(mesh, density=0.15)
+    >>> cpos = [(15, 3, 15), (0, 0, 0), (0, 0, 0)]
+    >>> vox.plot(scalars='MeshCells', show_edges=True, cpos=cpos)
+
+    Slice the voxel volume to view ``MeshCells``.
+
+    >>> slices = vox.slice_orthogonal()
+    >>> slices.plot(scalars='MeshCells', show_edges=True)
+
+    Create a voxel volume from unequal density dimensions and plot result.
+
+    >>> vox = pv.voxelize_volume(mesh, density=[0.15, 0.15, 0.5])
+    >>> vox.plot(scalars='MeshCells', show_edges=True, cpos=cpos)
+
+    Slice the unequal density voxel volume to view ``MeshCells``.
+
+    >>> slices = vox.slice_orthogonal()
+    >>> slices.plot(scalars='MeshCells', show_edges=True, cpos=cpos)
+
+    """
+    mesh = wrap(mesh)
+    if density is None:
+        density = mesh.length / 100
+    if isinstance(density, (int, float, np.number)):
+        density_x, density_y, density_z = [density] * 3
+    elif isinstance(density, (collections.abc.Sequence, np.ndarray)):
+        density_x, density_y, density_z = density
+    else:
+        raise TypeError(f'Invalid density {density!r}, expected number or array-like.')
+
+    # check and pre-process input mesh
+    surface = mesh.extract_geometry()  # filter preserves topology
+    if not surface.faces.size:
+        # we have a point cloud or an empty mesh
+        raise ValueError('Input mesh must have faces for voxelization.')
+    if not surface.is_all_triangles:
+        # reduce chance for artifacts, see gh-1743
+        surface.triangulate(inplace=True)
+
+    x_min, x_max, y_min, y_max, z_min, z_max = mesh.bounds
+    x = np.arange(x_min, x_max, density_x)
+    y = np.arange(y_min, y_max, density_y)
+    z = np.arange(z_min, z_max, density_z)
+
+    # Create a RectilinearGrid
+    voi = pyvista.RectilinearGrid(x, y, z)
+
+    # get part of the mesh within the mesh's bounding surface.
+    selection = voi.select_enclosed_points(surface, tolerance=0.0, check_surface=check_surface)
+    mask_vol = selection.point_data['SelectedPoints'].view(np.bool_)
+
+    # Get voxels that fall within input mesh boundaries
+    cell_ids = np.unique(voi.extract_points(np.argwhere(mask_vol))["vtkOriginalCellIds"])
+
+    # Create new element of grid where all cells _within_ mesh boundary are
+    # given new name 'MeshCells' and a discrete value of 1
+    voi['MeshCells'] = np.zeros(voi.n_cells)
+    voi['MeshCells'][cell_ids] = 1
+
+    # Create new element of grid where background cells
+    # given a new name 'BackgroundCells' and a discrete value of 0
+    voi['BackgroundCells'] = np.zeros(voi.n_cells)
+    voi['BackgroundCells'][0] = 0
```

Comment:
`'BackgroundCells'` is just 0 at every cell.  I recommend deleting it.  The needed information is all in `'MeshCells'`.

Reply:
Also I don't love the name of 'MeshCells', is there a better name?  I am terrible at naming though.

Reply:
I've chosen `MeshCells` because they are the cells that selected within the grid. I did have `SelectedCells` as a first iteration, but thought `MeshCells` made more sense becauase they are cells contained within and intersect the input Mesh.

Reply:
Where I get stuck is that the returned mesh already has cells.  I did a search for similar functions and found https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.DataSetFilters.select_enclosed_points.html#pyvista.DataSetFilters.select_enclosed_points

And now that I look at these, I'm not sure what the essential difference is other than this proposed filter presupplies a RectilinearGrid object to find enclosed points/cells.  That is why aren't we using that filter to generate this result?  Is it a difference between points and cells?

That method uses `SelectedPoints` for the name.

Reply:
This method is intended to return cells in a RectilinearGrid because that is what is commonly used in other software (e.g., fluid flow modelling in the subsurface). But also, and perhaps most importantly, this method generates a volume around the input mesh. I think having the ability to create a voxel volume that _encloses_ the mesh is a powerful feature to have built in. Extracting statistics of the interaction between the `SelectedCells` (if that is a more appropriate and agreed name) and the background cells has important utility. 

The filter you linked is used in this method to _select_ the cells contained within and that intersect the input mesh. This is the only way (at the moment) that I could discretise the voxel volume. As you mentioned before, using `GhostCells` is a good idea that would make this a fundamentally different method. 

I am happy to change the name to `SelectedCells` if that is what we think is clearer.

Reply:
@MatthewFlamm Are there any comments about this thread? If you don't have one, let's resolve this thread and merge PR.

Reply:
I still think `'BackgroundCells'` is unneeded. I don't love either `'MeshCells'` or `'SelectedCells'`.  But I'm also at a loss for a better name.  `'InsideMesh'`?

I'm only a bit concerned on the many seeming ways to do the same thing we are collecting.  This is in reference to understanding difference between this function and `select_enclosed_points`.  It isn't clear to me based on the examples and description how this function is different except that it provides a wrapper for creating a `RectilinearGrid`.  It is very likely these are different in some other way, but it doesn't come out in the description.

Reply:
I'm happy to remove `'BackgroundCells'`. These data can be extracted from the resultant `RectilinearGrid` the function creates. 

Naming is indeed difficult, but I think `'InsideMesh'` sounds good. It is descriptive enough for what the function does.

The function `select_enclosed_points` returns an `UnstructuredGrid`, which isn't that helpful in further modelling that require at least a structured grid of some kind. A `RectilinearGrid` is sufficient, and as we discussed above, fine to use for this function. You're right, this function essentially provides a wrapper to create a `RectilinearGrid`. I feel this is ok though, especially for people who may not be code saavy but want to use 'PyVista' for their visualisation/analysis work.

If the reviewers are ok with this line of thinking, then I'll make the following changes:
- Remove `BackgroundCells`
- Change the name to `InsideMesh`